xonsh/parsers/fstring_adaptor.py

```diff
@@ -0,0 +1,182 @@
+import re
+from xonsh.lazyasd import lazyobject
+from ast import parse as pyparse
+from xonsh import ast
+from xonsh.platform import PYTHON_VERSION_INFO
+
+
+@lazyobject
+def RE_FSTR_FIELD_WRAPPER():
+    if PYTHON_VERSION_INFO > (3, 8):
+        return re.compile(r"__xonsh__\.eval_fstring_field\((\d+)\)\s*[^=]")
+    else:
+        return re.compile(r"__xonsh__\.eval_fstring_field\((\d+)\)")
+
+
+if PYTHON_VERSION_INFO > (3, 8):
+
+    @lazyobject
+    def RE_FSTR_SELF_DOC_FIELD_WRAPPER():
+        return re.compile(r"(__xonsh__\.eval_fstring_field\((\d+)\)\s*)=")
+
+
+class FStringAdaptor:
+    def __init__(self, fstring, prefix, filename=None):
+        self.fstring = fstring
+        self.prefix = prefix
+        self.filename = filename
+        self.fields = {}
+        self.repl = ""
+        self.res = None
+
+    def _patch_special_syntax(self):
+        """Takes an fstring (and its prefix, ie f") that may contain
+        xonsh expressions as its field values and substitues them for
+        a call to __xonsh__.eval_fstring_field as needed.
+        """
+        prelen = len(self.prefix)
+        quote = self.fstring[prelen]
+        if self.fstring[prelen + 1] == quote:
+            quote *= 3
+        template = self.fstring[prelen + len(quote) : -len(quote)]
+        while True:
+            repl = self.prefix + quote + template + quote
+            try:
+                res = pyparse(repl)
+                break
+            except SyntaxError as e:
+                # The e.text attribute is expected to contain the failing
+                # expression, e.g. "($HOME)" for f"{$HOME}" string.
+                if e.text is None or e.text[0] != "(":
+                    raise
+                error_expr = e.text[1:-1]
+                epos = template.find(error_expr)
+                if epos < 0:
+                    raise
+            # We can olny get here in the case of handled SyntaxError.
+            # Patch the last error and start over.
+            xonsh_field = (error_expr, self.filename if self.filename else None)
+            field_id = id(xonsh_field)
+            self.fields[field_id] = xonsh_field
+            eval_field = f"__xonsh__.eval_fstring_field({field_id})"
+            template = template[:epos] + eval_field + template[epos + len(error_expr) :]
+
+        self.repl = repl
+        self.res = res.body[0].value
+
+    def _unpatch_strings(self):
+        """Reverts false-positive field matches within strings."""
+        reparse = False
+        for node in ast.walk(self.res):
+            if isinstance(node, ast.Constant) and isinstance(node.value, str):
+                value = node.value
+            elif isinstance(node, ast.Str):
+                value = node.s
+            else:
+                continue
+
+            match = RE_FSTR_FIELD_WRAPPER.search(value)
+            if match is None:
+                continue
+            try:
```

Comment:
please rewrite this to avoid the try-except. Something like:

```py
field = self.fields.pop(int(match.group(1)), None)
if field is None:
    continue
```

Reply:
Done.